i18n-calypso: Replace EventEmitter with simplified subscriber callbacks

aduth opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Part of ARC-129

## Proposed Changes

* Removes EventEmitter usage from `i18n-calypso` package, substituting a simplified subscriber implementation

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* Move away from outdated `EventEmitter` data approach toward conventions more conventional and compatible with React (e.g. `useEffect` subscribers)
* Reduce bundle size by eliminating `events` package in common dependency ([estimated 2kb gzipped](https://bundlephobia.com/package/events@3.3.0))
* Likely performance improvement due to lack of overhead of event emitter; we're not [handling multiple event types](https://github.com/browserify/events/blob/48e3d18659caf72d94d319871106f089bb40002d/events.js#L147), or [forwarding parameters](https://github.com/browserify/events/blob/48e3d18659caf72d94d319871106f089bb40002d/events.js#L121-L122), or [handling errors](https://github.com/browserify/events/blob/48e3d18659caf72d94d319871106f089bb40002d/events.js#L123).

On performance, I discovered that the `events` library does have a [surprisingly-performant subscriber removal behavior](https://github.com/browserify/events/blob/48e3d18659caf72d94d319871106f089bb40002d/events.js#L435-L439). Early iterations here included a "readable" `Array#filter` implementation (see a8fb2d473a593347e88cf7e82c0cc596aff21b2c), but considering that many of our localized components will remove event listeners once they're unmounted, it seemed worthwhile to micro-optimize. After several rounds of micro-benchmarking, I found a solution which performs even better (~38%) than `events`'s implementation, and ~850x better than a naive `Array#filter` implementation in a scenario with 2000 subscribers.

Separately, I had explored [an implementation which largely maintained the same API](https://github.com/Automattic/wp-calypso/compare/trunk...remove/i18n-events-dep-maintain-api) if we'd rather reduce the scope of the changes.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* The idea is that this should be drop-in compatible for localization, and should have no impact on the display and re-rendering of translation data. Changes can be emulated by switching languages in the dashboard.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

aduth committed:
Convert i18n events handling to subscriber model

aduth committed:
Add additional test coverage

aduth committed:
Microoptimize subscriber removal

aduth committed:
Update changelog for i18n-calypso

aduth committed:
Remove types for eliminated methods

yuliyan commented on packages/i18n-calypso/src/test/index.js:
```diff
@@ -266,6 +285,72 @@ describe( 'I18n', function () {
+	} );
+
+	describe( 'subscribe()', () => {
+		it( 'should call the callback whenever a change is emitted', () => {
```
Nit: This seems to be the same test as one of the tests from the [setLocale group](https://github.com/Automattic/wp-calypso/pull/103304/files#diff-c2019f571456ce7326d09b2bf702d03ebef06dd80efc81fb368abbf49b13c135R27).

aduth replied:
> Nit: This seems to be the same test as one of the tests from the [setLocale group](https://github.com/Automattic/wp-calypso/pull/103304/files#diff-c2019f571456ce7326d09b2bf702d03ebef06dd80efc81fb368abbf49b13c135R27).

Yeah, I struggled to think of how to test the effective behavior of `subscribe` without pairing it with something that emits a change. I'll probably remove one of these duplicates.

aduth replied:
> Nit: This seems to be the same test as one of the tests from the [setLocale group](https://github.com/Automattic/wp-calypso/pull/103304/files#diff-c2019f571456ce7326d09b2bf702d03ebef06dd80efc81fb368abbf49b13c135R27).

Removed duplicate test case in e8ea58b88d6.

yuliyan commented on packages/i18n-calypso/src/i18n.js:
```diff
@@ -191,16 +204,13 @@ I18N.prototype.geolocateCurrencySymbol = async function ( callback ) {
-	this.stateObserver.off( ...args );
+I18N.prototype.subscribe = function ( callback ) {
+	this.subscribers.push( callback );
+	return () => removeOne( this.subscribers, callback );
```
Would it make sense to use a `Set` for `subscribers` instead? That way, you can use `subscribers.delete()`, which would eliminate the need for the custom `removeOne` helper function and might also be better in terms of performance in some cases.

aduth replied:
> Would it make sense to use a `Set` for `subscribers` instead? That way, you can use `subscribers.delete()`, which would eliminate the need for the custom `removeOne` helper function and might also be better in terms of performance in some cases.

Yeah, that's a very valid question! I thought `Set` would be a good fit as well, but local benchmarking showed it to be slower at iterating compared to arrays. And while it's faster at removing than some of the naive array implementations using (using `Array#splice` or `Array#filter`), it's slower than the `removeOne` helper implemented here.

With 1000 subscribers:

Iteration:

```
SetSubscribers x 105,071 ops/sec ±0.49% (97 runs sampled)
ArraySubscribers x 135,548 ops/sec ±0.13% (101 runs sampled)
```

Removal: 

```
SetSubscribers x 133,243,391 ops/sec ±2.76% (87 runs sampled)
ArraySpliceSubscribers x 78,857,980 ops/sec ±1.94% (89 runs sampled)
RemoveOneSubscribers x 965,073 ops/sec ±0.46% (100 runs sampled)
```

<details><summary>Code</summary>
<pre><code>class SetSubscribers {
  subscribers = new Set();

  add(callback) {
    this.subscribers.add(callback);
    return () => this.subscribers.delete(callback);
  }

  emitChange() {
    this.subscribers.forEach((s) => s());
  }
}


class ArraySpliceSubscribers extends ArraySubscribers {
  subscribers = [];

  add(callback) {
    this.subscribers.push(callback);
    return () => this.subscribers.splice(this.subscribers.indexOf(callback), 1);
  }

  emitChange() {
    this.subscribers.forEach((s) => s());
  }
}


class RemoveOneSubscribers extends ArraySubscribers {
  subscribers = [];

  add(callback) {
    this.subscribers.push(callback);
    return () => removeOne(this.subscribers, callback);
  }
}</code></pre></details>

I had a conversation with ChatGPT about it on Friday, and it explained as follows, which made sense to me:

>Why is using Set slower for iterating and removing from an array?
>
>Using a `Set` to remove items from an array can be **slower** than using native array methods like `.filter()` or `splice()` in JavaScript for several key reasons:
>
>---
>
>### 🚫 1. **Sets aren't optimized for indexed iteration**
>
>* Arrays are optimized for **indexed, contiguous memory access**.
>* `Set` uses a **hash-based structure** with no numeric indexing.
>* Iterating a `Set` requires traversing an internal hash table (slower than array index access).
>
>---
>
>### 🐢 2. **Insertion into a Set isn't free**
>
>When you use a `Set` like this:
>
>```js
>const set = new Set(array);
>```
>
>* JavaScript needs to **hash every element** (even primitives get some internal tagging).
>* This adds upfront **O(n)** overhead just to create the `Set`.
>
>If you're doing:
>
>```js
>const set = new Set(array);
>set.delete(target);
>```
>
>* `delete()` is O(1), but **not noticeably faster than array-native methods** in real-world sizes.
>* And you still pay for creating the `Set`.
>
>---
>
>### 🧠 3. **JIT optimization favors arrays**
>
>V8 and other JS engines heavily optimize array operations like `.filter()` or `.splice()` for common use cases.
>
>* These optimizations rely on assumptions about data layout (dense arrays, no holes).
>* `Set` doesn’t benefit from these same performance paths because:
>
>  * It's unordered.
>  * It can't be accessed via integer indices.
>  * Its internal representation isn't as predictable.
>
>---
>
>### ⚡ Example of slower pattern:
>
>```js
>const set = new Set(array);
>set.delete(value);
>const newArray = Array.from(set); // extra allocation
>```
>
>This:
>
>1. Allocates a `Set` (O(n)),
>2. Performs a `delete()` (O(1)),
>3. Allocates a **new array** (O(n)).
>
>Compare to:
>
>```js
>array = array.filter(item => item !== value);
>```
>
>Which:
>
>* Scans once (O(n)),
>* Allocates one new array (O(n)),
>* Often benefits from JIT optimizations.
>
>---
>
>### ✅ Summary
>
>| Operation        | Arrays (`.filter`, `splice`) | `Set`                             |
>| ---------------- | ---------------------------- | --------------------------------- |
>| Iteration speed  | Fast (indexed, dense)        | Slower (hash traversal)           |
>| Memory layout    | Contiguous                   | Hashed                            |
>| JIT optimization | Excellent                    | Limited                           |
>| Best use case    | Fast filtering, ordered data | Fast existence checks, uniqueness |
>
>---

yuliyan replied:
@aduth, Thanks for the details! If I'm reading the benchmark data correctly - with a higher ops/second count indicating better performance - it looks like `Set.delete()` does the best with around 133 million ops/second, followed by `Array.splice()` at 79 million, and `removeOne` at about 1 million ops/second.

When testing locally, I noticed that `removeOne` began to show performance issues and getting worse exponentially at around 100,000 subscribers, especially when removing elements from the beginning of the array. I assume this is due to the higher number of reassignments involved.

That said, 100k subscribers is far beyond any real-world scenario I've seen. In my testing, most screens had fewer than 1,000 subscribed components (event handlers). So, I'm not sure this is something we need to worry about. And given that iteration performance is in favor using an Array, the overall performance likely balances out in practice.

aduth replied:
@yuliyan You're right, that's strange 🤔 Let me test that again and make sure I didn't make a typo or some other obvious error, because I could have sworn this was showing much better performance on Friday.

yuliyan replied:
Here's the combined (add, iterate, remove subscribers) script I ran for testing and the results:

```
Array add/remove x 10,360 ops/sec ±0.46% (99 runs sampled)
Set add/delete x 29,938 ops/sec ±0.66% (93 runs sampled)
```

<details>
<summary>Code:</summary>

~~~js
const Benchmark = require( 'benchmark' );

const subscribersCount = 1000;

const suite = new Benchmark.Suite();

// Test: Array add + remove
suite.add( 'Array add/remove', function () {
	const subscribers = [];
	const cb = [];

	function remove( arr, item ) {
		for ( let i = 0; i < arr.length; i++ ) {
			if ( arr[ i ] === item ) {
				for ( ; i < arr.length; i++ ) {
					arr[ i ] = arr[ i + 1 ];
				}
				arr.length--;
				return;
			}
		}
	}

	for ( let i = 0; i < subscribersCount; i++ ) {
		const fn = function () {};
		subscribers.push( fn );
		if ( i % 10 === 0 ) cb.push( fn );
	}

	subscribers.forEach( callback => callback() );

	for ( let i = 0; i < cb.length; i++ ) {
		remove( subscribers, cb[ i ] );
	}
} );

// Test: Set add + delete
suite.add( 'Set add/delete', function () {
	const subscribers = new Set();
	const cb = [];

	for ( let i = 0; i < subscribersCount; i++ ) {
		const fn = function () {};
		subscribers.add( fn );
		if ( i % 10 === 0 ) cb.push( fn );
	}

	subscribers.forEach( callback => callback() );

	for ( let i = 0; i < cb.length; i++ ) {
		subscribers.delete( cb[ i ] );
	}
} );

suite
	.on( 'cycle', function ( event ) {
		console.log( String( event.target ) );
	} )
	.run( { async: true } );

~~~
</details>

aduth replied:
Okay, so it was definitely slower than I recall. I think what happened is that I refactored an early iteration with the same general concept, not realizing that it made the code substantially slower. The original approach does seem to be very fast still:

```ts
function removeOne(array, item) {
  let knownIndex;
  for (let i = 0; i < array.length - 1; i++) {
    if (array[i] === item) {
      knownIndex = i;
    }

    if (knownIndex !== undefined) {
      array[i] = array[i + 1];
    }
  }
  array.pop();
}
```

```
SetSubscribers x 151,584,097 ops/sec ±3.15% (87 runs sampled)
RemoveOneSubscribers x 186,111,987 ops/sec ±3.27% (96 runs sampled)
```

(ran with callback as the 1001th entry of an array/set of 2001 subscribers)

<details><summary>Full code</summary>

~~~js
import Benchmark from "benchmark";

class SetSubscribers {
  subscribers = new Set();

  add(callback) {
    this.subscribers.add(callback);
    return () => this.subscribers.delete(callback);
  }

  emitChange() {
    this.subscribers.forEach((s) => s());
  }
}

class ArraySubscribers {
  subscribers = [];

  add(callback) {
    this.subscribers.push(callback);
    return () => {
      this.subscribers = this.subscribers.filter((s) => s !== callback);
    };
  }

  emitChange() {
    this.subscribers.forEach((s) => s());
  }
}

class ArraySpliceSubscribers extends ArraySubscribers {
  subscribers = [];

  add(callback) {
    this.subscribers.push(callback);
    return () => this.subscribers.splice(this.subscribers.indexOf(callback), 1);
  }

  emitChange() {
    this.subscribers.forEach((s) => s());
  }
}

function spliceShift(array, index) {
  for (; index < array.length - 1; index++) {
    array[index] = array[index + 1];
  }
  array.pop();
}

function spliceSwap(array, index) {
  array[index] = array[array.length - 1];
  array.pop();
}

function removeOne(array, item) {
  // for (let i = 0; i < array.length; i++) {
  //   if (array[i] === item) {
  //     for (; i < array.length - 1; i++) {
  //       array[i] = array[i + 1];
  //     }
  //     array.pop();
  //     return;
  //   }
  // }
  let knownIndex;
  for (let i = 0; i < array.length - 1; i++) {
    if (array[i] === item) {
      knownIndex = i;
    }

    if (knownIndex !== undefined) {
      array[i] = array[i + 1];
    }
  }
  array.pop();
}

class ArrayMicroSpliceSubscribers {
  subscribers = [];

  add(callback) {
    this.subscribers.push(callback);
    return () =>
      spliceShift(this.subscribers, this.subscribers.indexOf(callback));
  }
}

class UnimportantOrderSubscribers {
  subscribers = [];

  add(callback) {
    this.subscribers.push(callback);
    return () =>
      spliceSwap(this.subscribers, this.subscribers.indexOf(callback));
  }
}

class RemoveOneSubscribers {
  subscribers = [];

  add(callback) {
    this.subscribers.push(callback);
    return () => removeOne(this.subscribers, callback);
  }
}

const setSubscribers = new SetSubscribers();
const arraySubscribers = new ArraySubscribers();
const arraySpliceSubscribers = new ArraySpliceSubscribers();
const arrayMicroSpliceSubscribers = new ArrayMicroSpliceSubscribers();
const unimportantOrderSubscribers = new UnimportantOrderSubscribers();
const removeOneSubscribers = new RemoveOneSubscribers();
const suite = new Benchmark.Suite();

for (let i = 0; i < 1000; i++) {
  setSubscribers.add(() => {});
  arraySubscribers.add(() => {});
  arraySpliceSubscribers.add(() => {});
  arrayMicroSpliceSubscribers.add(() => {});
  unimportantOrderSubscribers.add(() => {});
  removeOneSubscribers.add(() => {});
}

const unsubscribeSetSubscribers = setSubscribers.add(() => {});
const unsubscribeArraySubscribers = arraySubscribers.add(() => {});
const unsubscribeArraySpliceSubscribers = arraySpliceSubscribers.add(() => {});
const unsubscribeArrayMicroSpliceSubscribers = arrayMicroSpliceSubscribers.add(
  () => {}
);
const unsubscribeUnimportantOrderSubscribers = unimportantOrderSubscribers.add(
  () => {}
);
const unsubscribeRemoveOneSubscribers = removeOneSubscribers.add(() => {});
for (let i = 0; i < 1000; i++) {
  setSubscribers.add(() => {});
  arraySubscribers.add(() => {});
  arraySpliceSubscribers.add(() => {});
  arrayMicroSpliceSubscribers.add(() => {});
  unimportantOrderSubscribers.add(() => {});
  removeOneSubscribers.add(() => {});
}

// suite
//   .add("SetSubscribers", () => {
//     setSubscribers.emitChange();
//   })
//   .add("ArraySubscribers", () => {
//     arraySubscribers.emitChange();
//   })
//   //   .add("ArraySpliceSubscribers", () => {
//   //     arraySpliceSubscribers.emitChange();
//   //   })
//   .on("cycle", (event) => {
//     console.log(String(event.target));
//   })
//   .on("complete", () => {
//     console.log("Fastest is " + suite.filter("fastest").map("name"));
//   })
//   .run({ async: true });

suite
  .add("SetSubscribers", () => {
    unsubscribeSetSubscribers();
  })
  .add("ArraySubscribers", () => {
    unsubscribeArraySubscribers();
  })
  .add("ArraySpliceSubscribers", () => {
    unsubscribeArraySpliceSubscribers();
  })

  .add("ArrayMicroSpliceSubscribers", () => {
    unsubscribeArrayMicroSpliceSubscribers();
  })
  .add("UnimportantOrderSubscribers", () => {
    unsubscribeUnimportantOrderSubscribers();
  })
  .add("RemoveOneSubscribers", () => {
    unsubscribeRemoveOneSubscribers();
  })
  .on("cycle", (event) => {
    console.log(String(event.target));
  })
  .on("complete", () => {
    console.log("Fastest is " + suite.filter("fastest").map("name"));
  })
  .run({ async: true });
~~~
</details>

Out of curiosity, I also checked a few different options for iterating sets and arrays. `Array#forEach` seems to be the fastest.

```
Set for...of x 1,712,260 ops/sec ±0.46% (100 runs sampled)
Array for...of x 3,939,224 ops/sec ±1.45% (98 runs sampled)
Set forEach x 269,467 ops/sec ±0.22% (98 runs sampled)
Array forEach x 4,190,848 ops/sec ±0.22% (100 runs sampled)
```

<details><summary>Full code</summary>

~~~js
import Benchmark from "benchmark";

const set = new Set();
const array = [];

for (let i = 0; i < 1000; i++) {
  set.add(i);
  array.push(i);
}

const suite = new Benchmark.Suite();

suite.add("Set for...of", () => {
  for (const value of set) {
    value;
  }
});

suite.add("Array for...of", () => {
  for (const value of array) {
    value;
  }
});

suite.add("Set forEach", () => {
  set.forEach((value) => {
    value;
  });
});

suite.add("Array forEach", () => {
  array.forEach((value) => {
    value;
  });
});

suite.on("cycle", (event) => {
  console.log(String(event.target));
});

suite.run({ async: true });
~~~
</details>

yuliyan replied:
I believe there's an issue with the this implementation of `removeOne`. The line `array.pop();` removes an item from the array regardless of whether a matching item was found (`knownIndex !== undefined`). As a result, it gradually empties the `subscribers` array after repeated calls to `unsubscribeRemoveOneSubscribers`, the array becomes empty. This might give the false impression of improved performance, but it's actually due to the subscribers array being depleted.

On a separate note, I think the test suites might not be entirely accurate. The `unsubscribe*` callbacks only remove an element during the first call and subsequent calls simply iterate over the array (or set) without actually removing anything. This could lead to misleading benchmarks or test results, since the cost of deletion only happens once.

To ensure that each test run performs a proper unsubscribe operation, tests should be like:

<details>
<summary>Test code</summary>

~~~js
function createMultipleUnsubscribers(unsubscribeFactory, numberOfUnsubscribers) {
	const unsubscribeFunctions = [];
	for (let i = 0; i < numberOfUnsubscribers; i++) {
		unsubscribeFunctions.push(unsubscribeFactory());
	}

	return () => unsubscribeFunctions.forEach(unsubscribe => unsubscribe());
}

suite
	.add('SetSubscribers', () => {
		const setSubscribers = new SetSubscribers();

		for (let i = 0; i < 400; i++) {
		setSubscribers.add(() => {});
	}

	const unsubscribeSetSubscribers = createMultipleUnsubscribers( () => setSubscribers.add(() => {}), 200);

		for (let i = 0; i < 400; i++) {
		setSubscribers.add(() => {});
	};

		unsubscribeSetSubscribers();
	})
	.add('RemoveOneSubscribers', () => {
		const removeOneSubscribers = new RemoveOneSubscribers();

		for (let i = 0; i < 400; i++) {
		removeOneSubscribers.add(() => {});
	}

	const unsubscribeRemoveOneSubscribers = createMultipleUnsubscribers( () => removeOneSubscribers.add(() => {}), 200);

		for (let i = 0; i < 400; i++) {
		removeOneSubscribers.add(() => {});
	}

		unsubscribeRemoveOneSubscribers();
	})
	.on('cycle', (event) => {
		console.log(String(event.target));
	})
	.on('complete', () => {
		console.log('Fastest is ' + suite.filter('fastest').map('name'));
	})
	.run({ async: true });
~~~
</details>

tyxla replied:
This is definitely a fun convo so far 😅 My personal recommendation here would be to use whatever is available natively, since browsers tend to micro-optimize these to a level we often can't even comprehend. With that in mind, I'd prefer @yuliyan's suggestion of going with a `Set` unless we have a really good reason not to.

tyxla replied:
It's also worth keeping in mind that those benchmarks aren't very reliable and can greatly vary based on a bunch of factors. I can go into detail if needed, but I bet @sgomes has a snippet to explain that since he's done it thousands of times 😅

sgomes replied:
Hey @tyxla ! 👋  I don't have a snippet, but I might create one going forward 😁

It's true that micro-benchmarks on the web are generally a bad idea. That's because we use them as a proxy for performance within a given application or system (which is what we actually care about), but they rarely represent the system accurately, for several reasons.

First of all, JS engines are complex, and often feature various tiers of optimisation. The first time you run a piece of code it might be interpreted, then it goes through a light optimisation, then after enough runs it gets fully optimised into something closer to native CPU instructions. There are a bunch of heuristics in this process, each engine is different, and everything changes as new versions are released, so it's effectively impossible to know when or even whether optimisations will kick in during "normal" usage of a website. Which means that if you run a piece of code a whole bunch of times, you're likely going to enter that highest level of optimisation, but your full system might not. So you might be measuring the wrong thing.

Then, there's the issue of size and locality. Micro-benchmarks tend to be small, which means that the code is more likely to hang around in fast caches, whereas that's not a guarantee for a similar piece of code that's part of a larger system. Particularly if it doesn't run as often or as sequentially as you'd expect.

There's also the issue of de-optimisations, and what I'd call "over-optimisations" in the micro-benchmark. When we reach the higher tiers of optimisation, things like data type really start to matter. The browser doesn't know a variable or a parameter's data type, so at some point it has to guess; this means that if a function gets run a bunch of times with an integer as its parameter, the browser might figure it's worth optimising to assume it always takes an integer. But then if it later gets a floating point number instead (or worse, a string), the optimisation will fail (what we'd call a de-optimisation) and the browser will fall back to a lower, slower optimisation tier to be able to handle both data types. This means that if you're not careful, your micro-benchmark could end up with a higher level of optimisation than what the full system might be able to achieve, because it's using a much stricter set of constraints that what the full system has to deal with.

And finally, there's the issue of necessity; is the micro-benchmark necessary at all in the first place? Have we identified this code as a hot path that needs to be optimised? Even if the micro-benchmark is representative and accurate, it might not be worth it if we're only saving a millisecond or a fraction thereof. Simple, readable, non-surprising code should be preferred, and only turned into something more complex if a performance problem is identified and measured (and the fix measured to be better). In general, there's no point optimising a data structure or function to the extreme, if it is then e.g. being run as part of a poor algorithm with exponential growth (which should be the first thing getting fixed), or only being run occasionally. Always remember Donald Knuth's words: "Premature optimization is the root of all evil".

Hope that clarifies why I tend to stay away from micro-benchmarks on the web, and why I'd encourage others to do the same!

If you do come across a performance problem you're struggling with, please give me a poke, and I'd be happy to help identify what's happening and help with a fix 👍

aduth replied:
Appreciate y'all reigning this in, as fun as it is to benchmark here. I do think this is a hot path, and there's an interesting conflict in wanting to optimize for _both_ iteration (change notifications) and removal (removing subscribers on unmount), as well as developer readability.

In reflection, I think the `Set` operations should be _good enough_, or at least good enough to try to unblock this conversation and move forward, and we can revisit performance if it becomes a bottleneck.

aduth replied:
Changed to `Set` in d8589a6e485

aduth replied:
I also did a bit of real-world debugging, and it looks like number-of-subscribers fluctuates between ~50 (`/log-in`) and ~300 (`/sites`), and I only saw a total of 2 `emitChange` calls. Probably not worth stressing too much over micro-optimizations.

yuliyan reviewed (approved):
The changes appear to be working well, and the rationale behind them makes sense. Regarding the suggestion to use a `Set` instead of an array for the subscribers - I haven't done a deep dive into the performance implications, so feel free to disregard that if using an array makes more sense in this context.
LGTM 👍

tyxla reviewed (commented):
Thanks for working on it @aduth 🙌 

Happy to see one more `EventEmitter` down. So few remaining!

The only thing I'm hesitant about is what @yuliyan already recognized with how we're cleaning up the subscribers.

jsnajdr commented on client/lib/translator-jumpstart/index.js:
```diff
@@ -328,6 +328,6 @@ export function trackTranslatorStatus( isTranslatorEnabled ) {
 
 // re-initialize when new locale data is loaded
-i18n.on( 'change', communityTranslatorJumpstart.init.bind( communityTranslatorJumpstart ) );
+i18n.subscribe( communityTranslatorJumpstart.init.bind( communityTranslatorJumpstart ) );
```
The call of the `init` method, does it even work? The method takes two args, `user` and `isUserSettingsReady`. And returns early if `user` is null-ish. But the `change` event listener was not called with any arguments, is it? And neither is the new `subscribe` callback.

aduth replied:
> The call of the `init` method, does it even work? The method takes two args, `user` and `isUserSettingsReady`. And returns early if `user` is null-ish. But the `change` event listener was not called with any arguments, is it? And neither is the new `subscribe` callback.

I'll check on this tomorrow 👍

aduth replied:
> > The call of the `init` method, does it even work? The method takes two args, `user` and `isUserSettingsReady`. And returns early if `user` is null-ish. But the `change` event listener was not called with any arguments, is it? And neither is the new `subscribe` callback.
> 
> I'll check on this tomorrow 👍

Update: You're correct that the initialization triggered by this subscriber always short-circuits at [the user check](https://github.com/Automattic/wp-calypso/blob/376526141529a79a53a966dec6d7cc2e38c1b137/client/lib/translator-jumpstart/index.js#L137-L141). This subscriber is very old (originates in pre-OSS version), at which time [the `init` function didn't take arguments](https://github.com/Automattic/wp-calypso/blob/53bde8c3f9c4668e62dbd2c670aaacc2bd820427/client/lib/translator-jumpstart/index.js#L143). Additionally, I don't think we should rely on a side-effect like this when importing a module, since it isn't compatible with tree-shaking. That being said, the subscriber _is_ being added.

Instead, I think [the initialization is happening in the `TranslatorLauncher` component](https://github.com/Automattic/wp-calypso/blob/376526141529a79a53a966dec6d7cc2e38c1b137/client/layout/community-translator/launcher.jsx#L36-L45), which passes the user arguments correctly. It also appears to reinitialize many times when the component re-renders. 

I'm not entirely familiar with how this _should_ be working (e.g. in response to asynchronous loads or language changes), but I think it's fair to say that the subscriber is not doing anything in its current form and can be removed.

aduth replied:
I'm going to keep this in as-is to keep it functionally equivalent, and will follow-up with a pull request proposing to remove the ineffective `subscribe` call.

ciampo commented on packages/i18n-calypso/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+## Trunk
+
+- Breaking: Emitter object replaced with subscriber
+  - Before: `i18n.on( 'change', callback ); i18n.off( 'change', callback );`
+  - After: `const unsubscribe = i18n.subscribe( callback ); unsubscribe();`
+
```
Is it necessary to introduce new APIs? And if it is, could we keep the `on` and `off` APIs around for back compact (while changing their internal implementation)?

aduth replied:
> Is it necessary to introduce new APIs? And if it is, could we keep the `on` and `off` APIs around for back compact (while changing their internal implementation)?

It's not strictly necessary. I made a brief mention of it in my original comment, with an alternative branch implementation that looks a lot more like what you suggested:

>Separately, I had explored [an implementation which largely maintained the same API](https://github.com/Automattic/wp-calypso/compare/trunk...remove/i18n-events-dep-maintain-api) if we'd rather reduce the scope of the changes.

My main reason for proposing this as-is as opposed to the alternative is:

- There may be some light overhead in maintaining the translation layer, particularly one that's only used for compatibility's sake and not one we think is an ideal representation of the intended behavior
- It may rather misleadingly suggest we support other events than just `change`.

I had also checked impact in looking for where we might be using this package outside Calypso. As best I can tell, we're only using this in Calypso and in code distributables compiled from Calypso (`apps/`).

jsnajdr commented on packages/i18n-calypso/src/i18n.js:
```diff
@@ -191,16 +204,13 @@ I18N.prototype.geolocateCurrencySymbol = async function ( callback ) {
-I18N.prototype.emit = function ( ...args ) {
-	this.stateObserver.emit( ...args );
+I18N.prototype.emitChange = function () {
+	this.subscribers.forEach( ( callback ) => callback() );
```
`emitChange` doesn't really need to be a public method. Can we avoid the publicity somehow?

aduth replied:
> `emitChange` doesn't really need to be a public method. Can we avoid the publicity somehow?

This could be a good candidate for converting to a `class` and using [private class methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_properties)?

As a compromise, it might be possible for me to effectively hide from the public API by excluding from the TypeScript types, even if technically accessible.

aduth replied:
> `emitChange` doesn't really need to be a public method. Can we avoid the publicity somehow?

Removed from types API in d29692b14dd.

Technically it's still accessible, but relying on TypeScript to avoid surfacing this and flagging unintended usage.

jsnajdr commented on packages/i18n-calypso/src/i18n.js:
```diff
@@ -27,6 +26,24 @@ const translationLookup = [
+			return;
+		}
+	}
+}
```
I'm suprised that this is faster than `.splice`, because this shift-and-pop operation is _exactly_ what `.splice` is supposed to do. `.splice` also returns an array of deleted elements -- that's the only reason why it should be slower, because we don't make any use of the return value.

I also checked the C++ `EventTarget` implementation and the `removeEventListener` implementation is very similar to this. It uses `std::vector` for the listener list, and the `.erase` method for removal. I.e., a contiguous array of items in memory, and `erase` shifts all remaining elements, an O(n) operation. If order of listeners didn't matter (it does, it's prescribed by standard), you could do an O(1) swap-and-pop.

aduth replied:
> I'm suprised that this is faster than `.splice`, because this shift-and-pop operation is _exactly_ what `.splice` is supposed to do. `.splice` also returns an array of deleted elements -- that's the only reason why it should be slower, because we don't make any use of the return value.
> 
> I also checked the C++ `EventTarget` implementation and the `removeEventListener` implementation is very similar to this. It uses `std::vector` for the listener list, and the `.erase` method for removal. I.e., a contiguous array of items in memory, and `erase` shifts all remaining elements, an O(n) operation. If order of listeners didn't matter (it does, it's prescribed by standard), you could do an O(1) swap-and-pop.

There's a bit more discussion at https://github.com/Automattic/wp-calypso/pull/103304#discussion_r2084955114 . This current implementation is actually slower than ideal, but there are versions of it that can be very fast.

It's all a bit moot since I'll be switching this to a simpler `Set` implementation, but I think a lot of the differences can be attributable to V8 internal optimizations.

I actually did try an iteration with the "swap-and-pop" approach you mentionnend but it benchmarked quite slow.

```ts
function removeOne(array, item) {
  const index = array.indexOf(item);
  if (index !== -1) {
    array[index] = array[array.length - 1];
    array.pop();
  }
}
```

I had also conversed with ChatGPT about this counter-intuitive result on Friday, which described it this way:

> Why does it seem to not perform as well in my benchmarking?
> 
> Here's my script:
> 
> [...]
> 
> Great benchmarking setup — and **excellent catch** on the counterintuitive result. Let’s dig into **why `spliceShift` is outperforming `spliceSwap`** in your > benchmark, even though in theory `spliceSwap` does less work.
> 
> ---
> 
> ### 🔍 The surprising performance difference is likely due to **array ordering assumptions and V8's optimization behavior**.
> 
> ---
> 
> ## 🚨 The Key Problem: **`spliceSwap` breaks array order**
> 
> When you do this:
> 
> ```js
> array[index] = array[array.length - 1];
> array.pop();
> ```
> 
> You're mutating the array in a way that **violates the “ordered” invariant** V8 often expects for fast-element arrays (a.k.a. "Smi" or "packed" arrays). > Specifically:
> 
> 1. **You copy the last element to the middle of the array.**
> 2. Then `pop()` removes the last.
> 
> From V8’s perspective:
> 
> * The array now has **out-of-order elements**.
> * It may change the internal representation from **"packed" to "holey"**.
> * This disables key JIT optimizations for array access.
> 
> ---
> 
> ### ✅ In contrast: `spliceShift` maintains order
> 
> ```js
> for (; index < array.length - 1; index++) {
>   array[index] = array[index + 1];
> }
> array.pop();
> ```
> 
> This:
> 
> * Shifts all items left by one index.
> * Keeps the array densely packed and ordered.
> * Aligns with how V8 internally expects array memory to behave.
> * Gets **super-optimized** in tight loops with predictable memory access.
> 
> ---
> 
> ### 🧪 Proof: Inspect element kinds (optional deep dive)
> 
> In Node.js, V8 internally uses **element kinds** to optimize arrays:
> 
> * `PACKED_ELEMENTS` – fast
> * `HOLEY_ELEMENTS` – slower
> * `DICTIONARY_ELEMENTS` – slowest
> 
> When you do something “weird” (like overwrite a middle index with a non-sequential value), it can degrade your array's internal structure.
> 
> ---
> 
> ### 🛠 Suggestions
> 
> #### 1. Benchmark `spliceSwap` with a smaller array (e.g., 10 items)
> 
> To isolate the actual cost of the swap vs. the shift.
> 
> #### 2. Check perf with primitive values instead of functions
> 
> Function objects may introduce GC or pointer indirection variance.
> 
> #### 3. Benchmark with V8 flags for diagnostics
> 
> You can run Node with:
> 
> ```bash
> node --trace-opt --trace-deopt benchmark.js
> ```
> 
> To see what gets optimized or deoptimized.
> 
> ---
> 
> ### ✅ Conclusion
> 
> * `spliceSwap` is **theoretically faster**, but it **violates performance assumptions** V8 relies on for fast array access.
> * `spliceShift`, though seemingly "heavier", stays in the **hot path** of JIT-optimized array code.
> 
> In real-world usage, **`spliceSwap` is only faster when array order doesn't matter AND the array is short-lived or already disordered**.
> 
> ---

aduth replied:
Changed implementation to `Set`-based in d8589a6e485. See discussion at https://github.com/Automattic/wp-calypso/pull/103304#discussion_r2084633829

jsnajdr reviewed (approved):
Nice discussion, but I'm afraid we're already deep in the "good enough" territory where all the proposed alternatives are perfectly fine.

aduth committed:
Use Set as internal representation for subscribers

Facilitate deletion by reference

See: https://github.com/Automattic/wp-calypso/pull/103304#discussion_r2084633829

aduth committed:
Omit emitChange from public types API

aduth committed:
Remove duplicative test on subscriber calls

Since we're interested in subscriber behavior in response to certain actions (e.g. setting locale), not relevant to test here, and duplicative.

See: https://github.com/Automattic/wp-calypso/pull/103304#discussion_r2084633300
